File: 2021/14/python/main.py
# Day 14

with open('../input.txt') as f:
    template = f.readline().strip()
    f.readline()

    rules = dict([l.strip().split(" -> ") for l in f.readlines()])

memo = {}

def expand(pair, steps):
    if (pair, steps) in memo:
        return memo[(pair, steps)].copy()

    if steps == 0:
        return {}

    middle = rules[pair]

    result = expand(pair[0] + middle, steps-1)
    result_right = expand(middle + pair[1], steps-1)

    for k,v in result_right.items():
        result[k] = result.get(k, 0) + v

    result[middle] = result.get(middle, 0) + 1

    memo[(pair, steps)] = result.copy()

    return result

def get_counts(steps):
    counts = {}
    for c in template:
        counts[c] = counts.get(c, 0) + 1

    for i in range(len(template)-1):
        pair = template[i:i+2]
        result = expand(pair, steps)

        for k,v in result.items():
            counts[k] = counts.get(k, 0) + v

    counts_list = sorted(counts.items(), key=lambda p: p[1])
    return counts_list[-1][1] - counts_list[0][1]

print("Part 1:", get_counts(10))
print("Part 2:", get_counts(40))

File: 2021/14/rust/src/main.rs
use cached::proc_macro::cached;
use std::collections::HashMap;
use std::error::Error;
use std::fs::read_to_string;

static mut RULES: Option<HashMap<String, String>> = None;

#[cached]
fn expand(pair: String, steps: u8) -> HashMap<String, u64> {
    if steps == 0 {
        return HashMap::new();
    }

    unsafe {
        match &RULES {
            Some(r) => {
                let middle_char = &r[&pair];

                let mut left_pair = pair[0..1].to_string();
                left_pair.push_str(&middle_char);
                let mut result = expand(left_pair, steps - 1);

                let mut right_pair = String::from(middle_char);
                right_pair.push_str(&pair[1..2]);
                let result_right = expand(right_pair, steps - 1);

                for (k, v) in result_right {
                    *result.entry(k).or_default() += v;
                }

                *result.entry(middle_char.to_string()).or_default() += 1;
                result
            }
            None => HashMap::new(),
        }
    }
}

fn get_counts(steps: u8, template: &str) -> u64 {
    let mut counts: HashMap<String, u64> = HashMap::new();
    for i in 0..template.len() {
        let c = &template[i..i + 1];
        *counts.entry(c.to_string()).or_default() += 1;

        if i < (template.len() - 1) {
            let pair = &template[i..i + 2];
            let result = expand(String::from(pair), steps);

            for (k, v) in result {
                *counts.entry(k.to_string()).or_default() += v;
            }
        }
    }

    let min = counts.values().min().unwrap();
    let max = counts.values().max().unwrap();

    max - min
}

fn parse_line(line: &&str) -> (String, String) {
    let tuple = line.split_once(" -> ").unwrap();

    (tuple.0.to_string(), tuple.1.to_string())
}

fn main() -> Result<(), Box<dyn Error>> {
    let file = read_to_string("../input.txt").unwrap();
    let lines = file.lines().collect::<Vec<&str>>();

    let template = lines[0];

    unsafe {
        RULES = Some(lines[2..].iter().map(parse_line).collect());
        println!("Part 1: {}", get_counts(10, template));
        println!("Part 2: {}", get_counts(40, template));
    }

    Ok(())
}
